oak-run/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreIterator.java

```diff
@@ -19,35 +19,50 @@
+
 import java.util.Iterator;
 import java.util.Set;
 
-import com.google.common.collect.AbstractIterator;
 import org.apache.jackrabbit.oak.index.indexer.document.NodeStateEntry;
-import org.apache.jackrabbit.oak.index.indexer.document.flatfile.FlatFileBufferLinkedList.NodeIterator;
+import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.FlatFileBufferLinkedList;
+import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.NodeStateEntryList;
+import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.PersistedLinkedList;
+import org.apache.jackrabbit.oak.spi.blob.BlobStore;
 import org.apache.jackrabbit.oak.spi.state.NodeState;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import static com.google.common.collect.Iterators.concat;
-import static com.google.common.collect.Iterators.singletonIterator;
+import com.google.common.collect.AbstractIterator;
 
 class FlatFileStoreIterator extends AbstractIterator<NodeStateEntry> implements Iterator<NodeStateEntry> {
     private final Logger log = LoggerFactory.getLogger(getClass());
     private final Iterator<NodeStateEntry> baseItr;
-    private final FlatFileBufferLinkedList buffer;
+    private final NodeStateEntryList buffer;
     private NodeStateEntry current;
     private final Set<String> preferredPathElements;
     private int maxBufferSize;
     static final String BUFFER_MEM_LIMIT_CONFIG_NAME = "oak.indexer.memLimitInMB";
-    private static final int DEFAULT_BUFFER_MEM_LIMIT_IN_MB = 100;
 
-    public FlatFileStoreIterator(Iterator<NodeStateEntry> baseItr, Set<String> preferredPathElements) {
+    // by default, use the PersistedLinkedList
+    private static final int DEFAULT_BUFFER_MEM_LIMIT_IN_MB = 0;
+
+    public FlatFileStoreIterator(BlobStore blobStore, String fileName, Iterator<NodeStateEntry> baseItr, Set<String> preferredPathElements) {
+        this(blobStore, fileName, baseItr, preferredPathElements,
+                Integer.getInteger(BUFFER_MEM_LIMIT_CONFIG_NAME, DEFAULT_BUFFER_MEM_LIMIT_IN_MB));
+    }
+
+    public FlatFileStoreIterator(BlobStore blobStore, String fileName, Iterator<NodeStateEntry> baseItr, Set<String> preferredPathElements, int memLimitConfig) {
         this.baseItr = baseItr;
         this.preferredPathElements = preferredPathElements;
 
-        int memLimitConfig = Integer.getInteger(BUFFER_MEM_LIMIT_CONFIG_NAME, DEFAULT_BUFFER_MEM_LIMIT_IN_MB);
-        if (memLimitConfig < 0) {
+        if (memLimitConfig == 0) {
+            log.info("Using a key-value store buffer: {}", fileName);
+            NodeStateEntryReader reader = new NodeStateEntryReader(blobStore);
+            NodeStateEntryWriter writer = new NodeStateEntryWriter(blobStore);
+            this.buffer = new PersistedLinkedList(fileName, writer, reader);
+        } else if (memLimitConfig < 0) {
             log.info("Setting buffer memory limit unbounded", memLimitConfig);
```

Comment:
Sorry I'm afraid I don't understand the comment... We don't need to keep the config value of memLimitConfig as a field in this class, but we need to pass it to the FlatFileBufferLinkedList, it's larger than 0. Smaller than 0 means "no limit, but use FlatFileBufferLinkedList", and 0 means "don't use FlatFileBufferLinkedList but use the new PersistedLinkedList instead of the in-memory list".

Reply:
Sorry, the comment was not clear enough. The `memLimitConfig` is the `log.info` call is not used because there is no placeholder `{}` in the message. Probably we just need to use `log.info("Setting buffer memory limit unbounded")` without passing the param.

Reply:
Ah yes!